### PR TITLE
Configure qdevice on interactive mode

### PR DESCRIPTION
**Pull request by liangxin1300**
#### Configure qdevice on a running cluster
```Bash
15sp2-1:~ # crm cluster init -y
  Configuring csync2
  Generating csync2 shared key (this may take a while)...done
  csync2 checking files...done
  Configuring corosync
WARNING: Not configuring SBD (/etc/sysconfig/sbd left untouched).
WARNING: Hawk not installed - not configuring web management interface.
  Waiting for cluster..............done
  Loading initial cluster configuration
  Done (log saved to /var/log/crmsh/ha-cluster-bootstrap.log)
15sp2-1:~ # crm cluster init qdevice
Do you want to configure QDevice (y/n)? y
  HOST or IP of the QNetd server to be used []15sp2-3
  TCP PORT of QNetd server [5403]
  QNetd decision ALGORITHM(ffsplit/lms) [ffsplit]
  QNetd TIE_BREAKER(lowest/highest/valid node id) [lowest]
  Whether using TLS on QDevice/QNetd(on/off/required) [on]
  COMMAND to run with absolute path; For multiple commands, use ";" to separate []
  
Configure Qdevice/Qnetd:
  Update configuration...done
  Qdevice certification process...done
  Enable corosync-qdevice.service in cluster
  Restarting cluster service
  Waiting for cluster..............done
  Enable corosync-qnetd.service on 15sp2-3
  Starting corosync-qnetd.service on 15sp2-3
  Done (log saved to /var/log/crmsh/ha-cluster-bootstrap.log)
```
#### Configure qdevice from beginning
```Bash
15sp2-1:~ # crm cluster init
  Configuring csync2
csync2 is already configured - overwrite (y/n)? y
  Generating csync2 shared key (this may take a while)...done
  csync2 checking files...done
/etc/corosync/authkey already exists - overwrite (y/n)? y
/etc/corosync/corosync.conf already exists - overwrite (y/n)? y
  
Configure Corosync:
  This will configure the cluster messaging layer.  You will need
  to specify a network address over which to communicate (default
  is eth0's network, but you can use the network address of any
  active interface).

  IP or network address to bind to [192.168.122.193]
  Multicast address [239.197.122.212]
  Multicast port [5405]
/etc/pacemaker/authkey already exists - overwrite (y/n)? y
  
Configure SBD:
  If you have shared storage, for example a SAN or iSCSI target,
  you can use it avoid split-brain scenarios by configuring SBD.
  This requires a 1 MB partition, accessible to all nodes in the
  cluster.  The device path must be persistent and consistent
  across all nodes in the cluster, so /dev/disk/by-id/* devices
  are a good choice.  Note that all data on the partition you
  specify here will be destroyed.

Do you wish to use SBD (y/n)? n
WARNING: Not configuring SBD - STONITH will be disabled.
WARNING: Hawk not installed - not configuring web management interface.
  Waiting for cluster..............done
  Loading initial cluster configuration
  
Configure Administration IP Address:
  Optionally configure an administration virtual IP
  address. The purpose of this IP address is to
  provide a single IP that can be used to interact
  with the cluster, rather than using the IP address
  of any specific cluster node.

Do you wish to configure a virtual IP address (y/n)? n
Do you want to configure QDevice (y/n)? y
  HOST or IP of the QNetd server to be used []15sp2-3
  TCP PORT of QNetd server [5403]
  QNetd decision ALGORITHM(ffsplit/lms) [ffsplit]
  QNetd TIE_BREAKER(lowest/highest/valid node id) [lowest]
  Whether using TLS on QDevice/QNetd(on/off/required) [on]
  COMMAND to run with absolute path; For multiple commands, use ";" to separate []
  
Configure Qdevice/Qnetd:
  Update configuration...done
  Qdevice certification process...done
  Enable corosync-qdevice.service in cluster
  Restarting cluster service
  Waiting for cluster..............done
  Enable corosync-qnetd.service on 15sp2-3
  Starting corosync-qnetd.service on 15sp2-3
  Done (log saved to /var/log/crmsh/ha-cluster-bootstrap.log)
```

### CODE REVIEW

File: crmsh/bootstrap.py [hunk 1]
```diff
@@ -1865,15 +1865,42 @@ def evaluate_qdevice_quorum_effect(mode):
...
+        return
+    qnetd_addr = prompt_for_string("HOST or IP of the QNetd server to be used")
+    if not qnetd_addr:
+        error("Address of QNetd is required")
+    qdevice_port = prompt_for_string("TCP PORT of QNetd server", default=5403)
+    qdevice_algo = prompt_for_string("QNetd decision ALGORITHM(ffsplit/lms)", default="ffsplit")
+    qdevice_tie_breaker = prompt_for_string("QNetd TIE_BREAKER(lowest/highest/valid node id)", default="lowest")
+    qdevice_tls = prompt_for_string("Whether using TLS on QDevice/QNetd(on/off/required)", default="on")
+    qdevice_heuristics = prompt_for_string("COMMAND to run with absolute path; For multiple commands, use \";\" to separate")
+    qdevice_heuristics_mode = prompt_for_string("MODE of operation of heuristics(on/sync/off)", default="sync") if qdevice_heuristics else None
```
Comment (tomschr) [inline crmsh/bootstrap.py:1882]: I'd recommend to add a space before the parenthesis:

```suggestion
    qdevice_algo = prompt_for_string("QNetd decision ALGORITHM (ffsplit/lms)", default="ffsplit")
    qdevice_tie_breaker = prompt_for_string("QNetd TIE_BREAKER (lowest/highest/valid node id)", default="lowest")
    qdevice_tls = prompt_for_string("Whether using TLS on QDevice/QNetd (on/off/required)", default="on")
    qdevice_heuristics = prompt_for_string("COMMAND to run with absolute path; For multiple commands, use \";\" to separate")
    qdevice_heuristics_mode = prompt_for_string("MODE of operation of heuristics (on/sync/off)", default="sync") if qdevice_heuristics else None
```

File: crmsh/corosync.py [hunk 2]
```diff
@@ -282,9 +282,15 @@ def valid_attr(self):
         if not utils.valid_port(self.port):
             raise ValueError("invalid qdevice port range(1024 - 65535)")
 
+        if self.algo not in ["ffsplit", "lms"]:
```
Comment (tomschr) [inline crmsh/corosync.py:285]: It's a minor issue, so it's not that important. :wink: 
Code-wise, it's perfectly fine to use a list. However, I normally use a tuple here when I have a constant array of values and want to emphasize that this is something that isn't modified:

```suggestion
        if self.algo not in ("ffsplit", "lms"):
```

File: crmsh/corosync.py [hunk 3]
```diff
@@ -282,9 +282,15 @@ def valid_attr(self):
...
+            raise ValueError("invalid ALGORITHM choice: '{}' (choose from 'ffsplit', 'lms')".format(self.algo))
+
         if self.tie_breaker not in ["lowest", "highest"] and not utils.valid_nodeid(self.tie_breaker):
             raise ValueError("invalid qdevice tie_breaker(lowest/highest/valid_node_id)")
 
+        if self.tls not in ["on", "off", "required"]:
```
Comment (tomschr) [inline crmsh/corosync.py:291]: Same here:

```suggestion
        if self.tls not in ("on", "off", "required"):
```

File: test/unittests/test_bootstrap.py [hunk 4]
```diff
@@ -1382,6 +1382,51 @@ def test_init_qdevice(self, mock_status, mock_ssh, mock_qdevice_configured, mock
...
+        mock_qdevice.return_value = mock_qdevice_inst
+
+        bootstrap.configure_qdevice_interactive()
+        mock_confirm.assert_called_once_with("Do you want to configure QDevice?")
+        mock_prompt.assert_has_calls([
+            mock.call("HOST or IP of the QNetd server to be used"),
+            mock.call("TCP PORT of QNetd server", default=5403),
+            mock.call("QNetd decision ALGORITHM(ffsplit/lms)", default="ffsplit"),
+            mock.call("QNetd TIE_BREAKER(lowest/highest/valid node id)", default="lowest"),
+            mock.call("Whether using TLS on QDevice/QNetd(on/off/required)", default="on"),
```
Comment (tomschr) [inline test/unittests/test_bootstrap.py:1424]: If you use spaces in one of my previous comments, you need them here too:

```suggestion
            mock.call("QNetd decision ALGORITHM (ffsplit/lms)", default="ffsplit"),
            mock.call("QNetd TIE_BREAKER (lowest/highest/valid node id)", default="lowest"),
            mock.call("Whether using TLS on QDevice/QNetd (on/off/required)", default="on"),
```

**Review by tomschr — COMMENTED** [review]
@liangxin1300 Some minor issues. Hope they make sense.

File: crmsh/bootstrap.py (hunk 1 above)
Comment (liangxin1300) [reply crmsh/bootstrap.py]: Changed, thanks!
Also changed the same problem in options for argparse

File: crmsh/corosync.py (hunk 2 above)
Comment (liangxin1300) [reply crmsh/corosync.py]: Changed, thanks!

File: crmsh/corosync.py (hunk 3 above)
Comment (liangxin1300) [reply crmsh/corosync.py]: Changed

File: test/unittests/test_bootstrap.py (hunk 4 above)
Comment (liangxin1300) [reply test/unittests/test_bootstrap.py]: Changed

**Comment by liangxin1300** [comment]
> @liangxin1300 Some minor issues. Hope they make sense.

Thank you @tomschr !

**Comment by liangxin1300** [comment]
@gao-yan Would you mind to take a look at this PR before we move on?
Thank you!

**Review by gao-yan — APPROVED** [review]
Looks good to me. Thanks for the good work!

File: crmsh/bootstrap.py [hunk 5]
```diff
@@ -1865,15 +1865,42 @@ def evaluate_qdevice_quorum_effect(mode):
...
+    if _context.yes_to_all or not confirm("Do you want to configure QDevice?"):
+        return
+    qnetd_addr = prompt_for_string("HOST or IP of the QNetd server to be used")
+    if not qnetd_addr:
+        error("Address of QNetd is required")
+    qdevice_port = prompt_for_string("TCP PORT of QNetd server", default=5403)
+    qdevice_algo = prompt_for_string("QNetd decision ALGORITHM (ffsplit/lms)", default="ffsplit")
+    qdevice_tie_breaker = prompt_for_string("QNetd TIE_BREAKER (lowest/highest/valid node id)", default="lowest")
+    qdevice_tls = prompt_for_string("Whether using TLS on QDevice/QNetd (on/off/required)", default="on")
+    qdevice_heuristics = prompt_for_string("COMMAND to run with absolute path; For multiple commands, use \";\" to separate")
```
[the remark this answers is not in the document]
Comment (liangxin1300) [reply crmsh/bootstrap.py]: OK, changed, thanks

**Pull request merged**